lib/logsource.c

```diff
@@ -174,6 +174,95 @@ log_source_flow_control_suspend(LogSource *self)
+}
+
+void
+log_source_dynamic_window_update_statistics(LogSource *self)
+{
+  dynamic_window_stat_update(&self->dynamic_window, window_size_counter_get(&self->window_size, NULL));
+  msg_trace("Updating dynamic window statistic", evt_tag_int("avg window size",
+                                                             dynamic_window_stat_get_avg(&self->dynamic_window)));
+}
+
+static inline void
+_decrease_window(LogSource *self)
+{
+  gsize new_full_window_size = MAX(self->options->init_window_size, self->full_window_size / 2);
+  gsize subtrahend = self->full_window_size - new_full_window_size;
+
+  gsize empty_window = window_size_counter_get(&self->window_size, NULL);
+  if (empty_window <= subtrahend)
+    {
+      subtrahend = empty_window == 0 ? 0 : empty_window - 1;
+      new_full_window_size = self->full_window_size - subtrahend;
+    }
+
+  window_size_counter_sub(&self->window_size, subtrahend, NULL);
+
+  msg_trace("Decreasing dynamic window", evt_tag_int("previous_window", self->full_window_size),
+            evt_tag_int("new_window", new_full_window_size), log_pipe_location_tag(&self->super));
+
+  self->full_window_size = new_full_window_size;
+  dynamic_window_release(&self->dynamic_window, subtrahend);
+}
+
+static inline void
+_increase_window(LogSource *self)
+{
+  gsize offered_dynamic = dynamic_window_request(&self->dynamic_window, self->full_window_size);
+
+  msg_trace("Increasing dynamic window", evt_tag_int("previous_window", self->full_window_size),
+            evt_tag_int("new_window", self->full_window_size + offered_dynamic), log_pipe_location_tag(&self->super));
+
+  self->full_window_size += offered_dynamic;
+
+  gsize old_window_size = window_size_counter_add(&self->window_size, offered_dynamic, NULL);
+  if (old_window_size == 0 && offered_dynamic != 0)
+    log_source_wakeup(self);
+}
+
+static void
+_release_dynamic_window(LogSource *self)
+{
+  g_assert(self->ack_tracker == NULL);
+
+  gsize dynamic_part = self->full_window_size - self->options->init_window_size;
+  msg_trace("Releasing dynamic part of the window", evt_tag_int("dynamic_window_to_be_released", dynamic_part),
+            log_pipe_location_tag(&self->super));
+  self->full_window_size -= dynamic_part;
+  window_size_counter_sub(&self->window_size, dynamic_part, NULL);
+  dynamic_window_release(&self->dynamic_window, dynamic_part);
+}
+
+void
+log_source_dynamic_window_realloc(LogSource *self)
+{
+  /* it is safe to assume that the window size is not decremented while this function runs,
+   * only incrementation is possible by destination threads */
+
+  /* TODO: add abstraction for heuristics */
+  gsize free_avg = dynamic_window_stat_get_avg(&self->dynamic_window);
+  if (free_avg > self->full_window_size * (self->options->dynamic_window_decrease_threshold / 100.0f))
+    _decrease_window(self);
+
+  if (free_avg < self->full_window_size * (self->options->dynamic_window_increase_threshold / 100.0f))
+    _increase_window(self);
+
```

Comment:
I would try to avoid float arithmetic here, at least if this is a fast path, which I cannot judge at this point.

Reply:
Fortunately, this is not a fast path, but we can get rid of the floating-point arithmetic.

Reply:
Am I right to assume that the reason we don't need locking around window allocation/release is that we always do this in the main thread? It seems our timer is in the main thread. And we either call this path if working is FALSE or we defer this call until the fetch() function finishes. In any case, we'll only do dynamic window updates from the main thread, thus we don't need any locking.

Nice.

Reply:
Please note that the actual heuristic has problems and probably has to be extended/re-implemented.

Reply:
> Am I right to assume that the reason we don't need locking around window allocation/release is that we always do this in the main thread?

That is correct, I really liked this too. :)